### PR TITLE
🐛 Rewrite heading processing

### LINKED ISSUE #469: GitLab TOC support
Hello!  Your extension is very convenient, but there are a few issues with the TOC, generated with `markdown.extension.toc.githubCompatibility=true` when using with GitLab (v. 11.11.3-ee, the latest one)

The first issue is the same I described [here](https://github.com/thlorenz/doctoc/issues/165) for `doctoc`.  The others are specific for this extension (`doctoc` processes them correctly).

Here is my `test.md` with the generated TOC:

```
- [MikroTik hAP ac²](#mikrotik-hap-ac²)
  - [Switch + Access Point](#switch--access-point)
  - [Настраиваем Доступ к Сети](#Настраиваем-Доступ-к-Сети)

## MikroTik hAP ac²

The [hAP ac²](https://mikrotik.com/product/hap_ac2) is a Dual-concurrent Access Point, that provides Wifi coverage for 2.4 GHz and 5 GHz frequencies at the same time. Five 10/100/1000 Ethernet ports provide Gigabit connections for your wired devices, USB can be used for external storage or 4G/LTE modem, and device supports IPsec hardware acceleration.

### Switch + Access Point

### Настраиваем Доступ к Сети
```

1. The 1st entry contains a specific symbol `²`. I described a similar issue [here](https://github.com/thlorenz/doctoc/issues/165). The real GitLab link for that header is
```
https://mydomain.tld/myrepo/test.md#mikrotik-hap-ac
```

2. The 2nd entry. Your extension incorrectly processes the `+` sign.  The real GitLab link for that header is
```
https://mydomain.tld/myrepo/test.md#switch-access-point
```

3. The 3rd entry. This is a Cyrillic text. Your extension does not convert upper case characters to lower case. The correct TOC entry for that header (the working one) should be
```
  - [Настраиваем Доступ к Сети](#настраиваем-доступ-к-сети)
```
And the real GitLab link for that header is
```
https://mydomain.tld/myrepo/test.md#%D0%BD%D0%B0%D1%81%D1%82%D1%80%D0%B0%D0%B8%D0%B2%D0%B0%D0%B5%D0%BC-%D0%B4%D0%BE%D1%81%D1%82%D1%83%D0%BF-%D0%BA-%D1%81%D0%B5%D1%82%D0%B8
```
---
In fact, `doctoc` works more correctly, but your extension is much more convenient, and I love it a lot. I'd like you have corrected those issues. Thanks in advance!

### LINKED ISSUE #603: Table of content creation fails when code block is indented in ordered list
#### What is the problem?
Table of content creation fails with the following error, when code block is indented in ordered list:
```
Cannot read property '1' of null
```
#### How can I reproduce it?
VSCode: v1.41.1
OS: Linux
Markdown All in One: v2.7.0

Copy and paste the following markdown into a file, e.g. `test.md`:
```
# first header



## Second header

1. test line1
2. check if works as expected:
	```bash
	# comment
	ls
	```

```
then try to add TOC on line 3 and you will receive the error above.

#### Is there any error message in the console?
```
notificationsAlerts.ts:38 TypeError: Cannot read property '1' of null
	at /home/user/.vscode/extensions/yzhang.markdown-all-in-one-2.7.0/dist/extension.js:1
	at Array.map (<anonymous>)
	at h (/home/user/.vscode/extensions/yzhang.markdown-all-in-one-2.7.0/dist/extension.js:1)
	at /home/user/.vscode/extensions/yzhang.markdown-all-in-one-2.7.0/dist/extension.js:1
	at Generator.next (<anonymous>)
	at /home/user/.vscode/extensions/yzhang.markdown-all-in-one-2.7.0/dist/extension.js:1
	at new Promise (<anonymous>)
	at n (/home/user/.vscode/extensions/yzhang.markdown-all-in-one-2.7.0/dist/extension.js:1)
	at d (/home/user/.vscode/extensions/yzhang.markdown-all-in-one-2.7.0/dist/extension.js:1)
	at /home/user/.vscode/extensions/yzhang.markdown-all-in-one-2.7.0/dist/extension.js:1
	at Generator.next (<anonymous>)
	at /home/user/.vscode/extensions/yzhang.markdown-all-in-one-2.7.0/dist/extension.js:1
	at new Promise (<anonymous>)
	at n (/home/user/.vscode/extensions/yzhang.markdown-all-in-one-2.7.0/dist/extension.js:1)
	at c (/home/user/.vscode/extensions/yzhang.markdown-all-in-one-2.7.0/dist/extension.js:1)
	at v._executeContributedCommand (/opt/visual-studio-code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:538)
	at v.$executeContributedCommand (/opt/visual-studio-code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:539)
	at p._doInvokeHandler (/opt/visual-studio-code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:663)
	at p._invokeHandler (/opt/visual-studio-code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:662)
	at p._receiveRequest (/opt/visual-studio-code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:661)
	at p._receiveOneMessage (/opt/visual-studio-code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:660)
	at /opt/visual-studio-code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:658
	at l.fire (/opt/visual-studio-code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:46)
	at _.fire (/opt/visual-studio-code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:189)
	at /opt/visual-studio-code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:820
	at l.fire (/opt/visual-studio-code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:46)
	at _.fire (/opt/visual-studio-code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:189)
	at t.PersistentProtocol._receiveMessage (/opt/visual-studio-code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:193)
	at /opt/visual-studio-code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:190
	at l.fire (/opt/visual-studio-code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:46)
	at p.acceptChunk (/opt/visual-studio-code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:186)
	at /opt/visual-studio-code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:186
	at Socket.t (/opt/visual-studio-code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:195)
	at Socket.emit (events.js:200)
	at addChunk (_stream_readable.js:294)
	at readableAddChunk (_stream_readable.js:275)
	at Socket.Readable.push (_stream_readable.js:210)
	at Pipe.onStreamRead (internal/stream_base_commons.js:166)
```

### LINKED ISSUE #802: Table of contents: Heading anchor links for Azure DevOps Server 2019
ToC generation is a great feature, thank you for creating and maintaining it.

#### What is the problem?
My markdown file has headers with dots '.' inside, e.g. "# Interface Version 1.0". ToC generation with default settings generates an anchor link for this header, suppressing the dot:  "#interface-version-10". This is perfectly in line with the markdown spec, but unfortunately does not work in Azure Devops Server 2019.  I tried other slugify-modes than the default, but they all do not work in Azure Devops. What works in Azure Devops is keeping the dot: "#interface-version-1.0".

Is there any workaround known? If not, would it be possible to add a slugify-mode for Azure Devops 2019? 

#### How can I reproduce it?
* Setup a project in Azure Devops 2019
* Create a markdown file with at least on header having a dot '.' inside
* Generate the ToC
* Commit the code to Azure Devops, and preview the markdown file via the web interface.  

#### Is there any error message in the console?
n/a

### LINKED ISSUE #830: Table of contents not recognized if indented.
## Problem
A table of contents is not recognized as such if it is indented, e.g. within a `<details>` section.
(Code lens is missing and the update ToC command does not update it.)

## How to reproduce

Example file:

```markdown
# Test File

<details>
  <summary>Table of contents</summary>

  - [Test File](#test-file)
    - [SubHeading 1](#subheading-1)
    - [SubHeading 2](#subheading-2)
</details>

## SubHeading 1

## SubHeading 2
```

### LINKED ISSUE #831: Markdown Intellisense with multiple identical labels shows non-unique anchors
## Problem

See screenshot. Each "item" should have a unique anchor value: "item", "item-1", "item-2" etc. Intellisense just shows the "item" multiple times. The table of contents was generated with an extension just to show how the links should be; it's not relevant to the issue.

Note: the behavior is the same with editor.wordBasedSuggestions on or off. 

image

## How to reproduce

Just make a file that looks like the screenshot. And try to make an internal link to one of the item headers.

```
# item

# item

# item

[link]()
```

### LINKED ISSUE #853: "Create Table of Contents" fails when the file contains tab
This seems related to previous bugs but was not able to isolate the problem. Installed the latest dev build and still reproduces.

## Problem

I'm attempting to insert a ToC for the markdown here: 

https://bitbucket.org/ParadoxGBB/cinadmin/src/master/README.md

The template of the format is provided by BitBucket.org.

This might be due to the commenting style I have for my code samples (# is a comment in PowerShell), however all of those examples are indented so should have been covered by other fixes.

## How to reproduce

F1 + "Create Table of Contents"

## Error message in the console

Command 'Markdown All in One: Create Table of Contents' resulted in an error (Cannot read property '1' of null)

### PR DESCRIPTION
**Pull request by Lemmingh**
## Summary

* Fix #853, Fix #603: Perform leading tab-space conversion required by the CommonMark Spec in `getAllRootHeading()`. (Rewrite the whole function.)
* Fix #831: Expose proper slug in `IHeading`.
* Fix #830: Rewrite `detectTocRanges()` on top of markdown-it to eliminate various mysterious bugs.
* Fix #469: Align with GitLab 13.8.
* Correct `REGEX_FENCED_CODE_BLOCK`, `isInFencedCodeBlock()`.
* Correct/Adapt related functions.
* Correct `slugify()`.
* Correct GitHub slugify function.
* Correct Gitea (blackfriday) slugify function.
* Rewrite many fsPath/object-property-based dictionaries as Uri/Map-based, to mitigate various mysterious bugs.

### CODE REVIEW

File: src/toc.ts [hunk 1]
```diff
@@ -353,61 +430,196 @@ function loadTocConfig(editor: TextEditor): void {
...
+            isInToc: true,
+        };
+
+        // Identify ignored headings.
+        // We have trimmed trailing space or tab characters for every line in "transformations".
+
+        // If a parent heading has been omitted, also omit its children (subheadings).
+        if (respectProjectLevelOmit
+            && ignoredDepthBound !== undefined
+            && entry.level > ignoredDepthBound
```
Comment (yzhang-gh) [inline src/toc.ts:522]: What about a linebreak before `respectProj...`

File: src/toc.ts [hunk 2]
```diff
@@ -353,61 +430,196 @@ function loadTocConfig(editor: TextEditor): void {
...
-            const entry = {
-                level: matches[1].length,
-                text: matches[2].replace(/ #+ *$/, '').trim(),
-                lineNum: index,
-            };
-            return entry;
+            entry.isInToc = false;
+        }
+
+        if (respectMagicCommentOmit
```
Comment (yzhang-gh) [inline src/toc.ts:527]: also here

File: src/toc.ts [hunk 3]
```diff
@@ -470,25 +484,83 @@ export function getAllRootHeading(doc: TextDocument): IHeading[] {
...
+    doc,
+    respectMagicCommentOmit = false,
+    respectProjectLevelOmit = false,
+    slugifyMode = workspace.getConfiguration('markdown.extension.toc').get<SlugifyMode>('slugifyMode')!,
+}: {
+    doc: TextDocument;
+    respectMagicCommentOmit?: boolean;
+    respectProjectLevelOmit?: boolean;
+    slugifyMode?: SlugifyMode;
+}): ITocItem[] {
```
Comment (yzhang-gh) [inline src/toc.ts]: What is the purpose of passing a param object? type check?

**Review by yzhang-gh — COMMENTED** [review]
in progess

File: src/toc.ts [hunk 4]
```diff
@@ -163,7 +163,7 @@ function removeSectionNumbers() {
         return;
     }
     const doc = editor.document;
-    const toc = getAllRootHeading(editor.document);
+    const toc: readonly Readonly<IHeadingBase>[] = getAllRootHeading(doc, false, false);
```
Comment (yzhang-gh) [inline src/toc.ts:167]: Do we need default values for the last two parameters? It is not that intuitive from the name `getAllRootHeading`.

I can see that they were optional and you changed your mind later.

File: src/completion.ts [hunk 5]
```diff
@@ -559,20 +559,19 @@ class MdCompletionItemProvider implements CompletionItemProvider {
             const range = new Range(position.with({ character: startIndex + 1 }), endPosition);
 
             return new Promise((res, _) => {
-                const toc = getAllRootHeading(document);
+                const toc: readonly Readonly<IHeading>[] = getAllTocEntry({ doc: document, respectMagicCommentOmit: false, respectProjectLevelOmit: false });
 
-                const headingCompletions = toc.reduce((prev, curr) => {
-                    let item = new CompletionItem('#' + slugify(curr.rawContent), CompletionItemKind.Reference);
+                const headingCompletions = toc.map<CompletionItem>((heading) => {
```
Comment (yzhang-gh) [inline src/completion.ts:564]:
```diff
- (heading)
+ heading
```

File: src/util.ts [hunk 6]
```diff
@@ -21,19 +21,34 @@ export function isMdEditor(editor: TextEditor) {
...
+            token.type === 'fence'
+            && token.tag === 'code'
+            && token.map![0] <= lineIndex
+            && lineIndex < token.map![1]
+        ) {
+            return true;
+        }
+    }
+
+    return false;
```
Comment (yzhang-gh) [inline src/util.ts]: This looks good.

Just curious what is the effect on the performance (e.g. documents larger than our default `sizeLimit`). And also any corner case that the previous tricky regex way cannot handle?

File: src/test/suite/toc.test.ts [hunk 7]
```diff
@@ -772,17 +772,17 @@ suite("TOC.", () => {
                 '---',
                 '# Heading 1',
                 '## 1.2. Heading 1.1',
-                '2. Heading 2',
+                '2. Not Heading',
```
Comment (yzhang-gh) [inline src/test/suite/toc.test.ts:775]: 🤣

**Review by yzhang-gh — COMMENTED** [review]
Thanks for the great effort! ❤

One last question: should we force all the users to adopt the `[heading](<link>)`-style TOC? It will affect hundreds of thousands of people (creating a change to their git history). I would suggest not adding this when/where it is not necessary.

File: src/toc.ts (hunk 3 above)
Comment (Lemmingh) [reply src/toc.ts]: `getAllTocEntry()` accepts **a good number of optional parameters**. Then, I wanna make their order flexible, so that don't have to change callers every time making tiny changes to it.

I do expect **frequent** changes to the signature of `getAllTocEntry()`.

Comment (yzhang-gh) [reply src/toc.ts]: Fair enough. Or we can at least make `doc` positional and the rest as a "config" object, just like many VS Code APIs do.

File: src/util.ts (hunk 6 above)
Comment (Lemmingh) [reply src/util.ts]: > performance

* I did not experience notable lag when editing lists in a document with 58k characters (1k lines).
* markdown-it itself is already fast enough.

---

> case

A fenced code block can be deeply nested in other block structures. To recognize them correctly, **we have to parse the *whole* document**.

Example:

```markdown
> ~~~hello
> `world`
```

File: src/toc.ts (hunk 4 above)
Comment (yzhang-gh) [reply src/toc.ts:167]: Let me take a closer look tonight.

File: src/util.ts (hunk 6 above)
Comment (yzhang-gh) [reply src/util.ts]: 👌

If there aren't obvious disadvantages, of course we should use this more principle way. Many thanks!

**Comment by yzhang-gh** [comment]
I've made some changes (tried to improve the readability a little bit). Please correct me if I misunderstood something.

And in case you didn't notice a previous comment (see above), what is your opinion about, I would say, the breaking change: `[heading](<link>)`-style TOC?

**Comment by yzhang-gh** [comment]
As far as I can see, `getAllRootHeading(doc, rspMagicCmtOmit, rspProjLvlOmit)` has been used in
- `addSectionNumbers - (true, true)` and `.filter(i => i.isInToc ...)`
- `removeSectionNumbers - (false, false)`
- `getAllTocEntry - (false, false)`

And `getAllTocEntry` has been used in
- `generateTocText - (true, true)` and `.filter(i => i.isInToc ...)`
- `provideCompletionItems - (false, false)`

Looks like we can make `rspMagicCmtOmit, rspProjLvlOmit` (`getAllRootHeading`) defaults to `false`? (They are already `false` in `getAllTocEntry`.)

**Comment by yzhang-gh** [comment]
And a last thing about the naming: I see the point that you use `rootHeadings` to emphasize they have the real _absolute_ `level`s, while the TOC and section numbers use the _relative_ `level`s. IMHO, `heading` is enough for the absolute case and we already have many hints in the `generateTocText` function, like `const relativeLevel`. (It did take me a while to understand why it is called `root`.)

**Comment by Lemmingh** [comment]
> the `[heading](<link>)`-style TOC? It will affect hundreds of thousands of people

**If it's a big concern, I'll be willing to roll back**, and solve <https://github.com/yzhang-gh/vscode-markdown/issues/802#issuecomment-726296890> by another way (encode every character).

---

I made the design based on

* [**RFC 3986** section 2.3](https://www.rfc-editor.org/rfc/rfc3986#section-2.3)

  > For consistency, percent-encoded octets in the ranges of Unreserved Characters should not be created by URI producers ...

* [**CommonMark Spec 0.29** section 6.5](https://spec.commonmark.org/0.29/#link-destination)

  > A link destination consists of either ...

**Comment by yzhang-gh** [comment]
> roll back

I agree. It is good for the user experience.

> I made the design based on

I see. Thanks for the research 👍.

**Comment by Lemmingh** [comment]
> Looks like we can make `respectMagicCommentOmit, respectProjectLevelOmit` (`getAllRootHeading`) defaults to `false`?

Then, I can revert 35a228ea16714eaf50ba422ed6ce2178eab65ec0.

---

When first introducing them, I set the default value to `false`, based on the facts:

* When no value is passed, a parameter is `undefined`, thus, is **falsy**.
* In practice, an optional boolean parameter is usually given default value `false`.

Later, [the IntelliSense info](https://github.com/yzhang-gh/vscode-markdown/pull/877#discussion_r552419452) got me exhausted (It's horrible to read the hover info as a consumer), and reminded me of the nightmare of unreliable default value in some corner cases.

I was just scared. 😂

**Comment by yzhang-gh** [comment]
😄

**Comment by Lemmingh** [comment]
> It did take me a while to understand why it is called `root`.

https://github.com/yzhang-gh/vscode-markdown/blob/924492166e048d2e26ae135800a0571c55d71218/src/toc.ts#L435-L439

The "root" in `getAllRootHeading` refers to **the position in the document tree**.

Many block structures accept block structures as their content.

You can create some deeply nested things like:

```markdown
# Heading A

> + # Heading B ###
```

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE document SYSTEM "CommonMark.dtd">

<document xmlns="http://commonmark.org/xml/1.0">
  <heading level="1">
    <text>Heading A</text>
  </heading>
  <block_quote>
    <list type="bullet" tight="true">
      <item>
        <heading level="1">
          <text>Heading B</text>
        </heading>
      </item>
    </list>
  </block_quote>
</document>
```

---

https://github.com/yzhang-gh/vscode-markdown/blob/924492166e048d2e26ae135800a0571c55d71218/src/toc.ts#L14-L18
https://github.com/yzhang-gh/vscode-markdown/blob/924492166e048d2e26ae135800a0571c55d71218/src/typing/markdownSpec.d.ts#L23-L27

The "level" in `IHeading` refers to **the property of the heading** defined by the CommonMark Spec.

**Comment by yzhang-gh** [comment]
> The "root" in `getAllRootHeading` refers to **the position in the document tree**.
> 
> Many block structures accept block structures as their content.
> 
> You can create some deeply nested things like:
> 
> ```gfm
> # Heading A
> 
> > + # Heading B ###
> ```

But we never care about those non-root headings, right 😳? It is true that they are "headings" tags _in terms of the syntax tree_, but we won't say they are "headings" _in terms of a document/article_.

Also, it looks like we didn't make use of the syntax tree at all in this function. So I guess it would be better to move this "root" thing from the name to a piece of comment.

**Comment by Lemmingh** [comment]
> But we never care about those non-root headings

It's just our current architecture cannot handle them. (the same reason as [stated above](https://github.com/yzhang-gh/vscode-markdown/pull/877#discussion_r552446557))

**We indeed have to consider them!**

They play an important role in generating **slugs**. And implementations differ.

---

To solve all the problems, I'm privately prototyping version 4, where almost all features are powered by two language servers (one for symbol info and theming, the other for actions), and a plugin system is first class member.

I now get stuck constructing AST from markdown-it Tokens.

**Comment by yzhang-gh** [comment]
> > But we never care about those non-root headings
> 
> It's just our current architecture cannot handle them. (the same reason as [stated above](https://github.com/yzhang-gh/vscode-markdown/pull/877#discussion_r552446557))
> 
> **We indeed have to consider them!**
> 
> They play an important role in generating **slugs**. And implementations differ.

Probably I don't have enough context (about your plan). I was just saying it (assuming `heading` is `rootHeading`) is enough for the current feature set (TOC, section numbering). A so-called "heading" in a fenced code block is not a valid heading _in our features_ (not in the TOC, not involved in the section numbering, and not present in the reference links). Or are there other cases where a heading nested in some blocks should be considered in the current feature set?

**Comment by yzhang-gh** [comment]
> Or you plan to use the AST in the `getAllRootHeading` function soon/later?

Not sure whether we should do it in this PR, but #880 is an example supporting this.

**Comment by Lemmingh** [comment]
> I don't have enough context (about your plan).

I'll email you more details next month.

<details>

```typescript
export type ProviderResult<T> = T | Thenable<T>;

export interface IHeadingInfoProvider<T extends IHeadingInfo = IHeadingInfo> {

    readonly provideHeadingInfo: (document: vscode.TextDocument, options: IHeadingInfoOption, token: vscode.CancellationToken) => ProviderResult<T[]>;

    readonly resolveHeadingInfo: (items: T[], token: vscode.CancellationToken) => ProviderResult<Required<T>[]>;
}

export interface IHeadingInfoOption {
    documentMode: SlugifyMode;
    projectLevelOmittedHeading: [markdownSpec.MarkdownHeadingLevel, string][] | undefined;
    respectMagicCommentOmit: boolean;
}

export interface IHeadingInfo {

    location: vscode.Location;
    level: markdownSpec.MarkdownHeadingLevel;
    rawContent: string;

    canInToc?: boolean;
    slug?: string;
    visibleText?: string;
}
```

</details>

---

> Do you mean there are other cases where a heading nested in some blocks should be considered in the current feature set?

Yes. They play an important role in generating **slugs**. And I tend to regard the existence of `getAllRootHeading()` as a compromise. It's best to merge `getAllTocEntry()` and `getAllRootHeading()` as one method `getAllHeading()`, but the current architecture makes it too hard.

Render this example:

```markdown
# Heading A

> + # Heading B ###

# Heading B
```

It's evident that GitHub, GitLab, VS Code, Azure DevOps ... all **assign anchor ID to all headings** no matter where they are.

Then, you have to get all headings to calculate slugs correctly.

---

> Or you plan to use the AST in the `getAllRootHeading` function soon/later?

**Probably in 2022. NOT now.** I cannot make it within this month.

I tried to work directly on token stream. But as you can see in `detectTocRanges()`, it's frustrating. Then I decided to create something similar to Markdig, however, **markdown-it appears to erase lots of information, which is useless for rendering but critical for understanding a document**. So, stuck.

**Comment by yzhang-gh** [comment]
> And I tend to regard the existence of `getAllRootHeading()` as a compromise. It's best to merge `getAllTocEntry()` and `getAllRootHeading()` as one method `getAllHeading()`, but the current architecture makes it too hard.

Makes sense to me 👍

> It's evident that GitHub, GitLab, VS Code, Azure DevOps ... all **assign anchor ID to all headings** no matter where they are.

Interesting. I didn't know it before.

> > Or you plan to use the AST in the `getAllRootHeading` function soon/later?
> 
> **Probably in 2022. NOT now.** I cannot make it within this month.

That is fine.

BTW, my next big plan is to implement the "export as PDF" feature and that will bring us to the next major update (v4.0.0), although I am not sure when it will be done 🤣. Just take your (our) time.

File: src/util.ts [hunk 8]
```diff
@@ -332,21 +335,19 @@ const slugifyMethods: { readonly [mode in SlugifyMode]: (rawContent: string) =>
     /**
      * GitLab
      */
-    "gitlab": (slug: string): string => {
+    [SlugifyMode.GitLab]: (slug: string): string => {
```
Comment (yzhang-gh) [inline src/util.ts:338]: Did you forget to do it for other slugify modes?

Comment (Lemmingh) [reply src/util.ts:338]: No. I just did not want to create a too long diff.

Then, let's make them consistent.

**Comment by yzhang-gh** [comment]
🍻
